pkg/ovsdb/transact.go

```diff
@@ -1556,52 +1513,77 @@ func preWait(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.Op
 	return etcdGetByWhere(txn, ovsOp, ovsResult)
 }
 
-func doWaitDummy(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
-	return nil
-}
-
+/* wait */
 func doWait(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
-	rows := []map[string]interface{}{}
+	if ovsOp.Table == nil {
+		klog.Errorf("missing table parameter")
+		return errors.New(E_CONSTRAINT_VIOLATION)
+	}
+
+	if ovsOp.Rows == nil {
+		klog.Errorf("missing rows parameter")
+		return errors.New(E_CONSTRAINT_VIOLATION)
+	}
+
+	if ovsOp.Columns == nil {
+		klog.Errorf("missing columns parameter")
+		return errors.New(E_CONSTRAINT_VIOLATION)
+	}
+
+	if ovsOp.Until == nil {
+		klog.Errorf("missing until parameter")
+		return errors.New(E_CONSTRAINT_VIOLATION)
+	}
+
 	tableSchema, err := txn.schemas.LookupTable(txn.request.DBName, *ovsOp.Table)
 	if err != nil {
+		klog.Errorf("%s", err)
 		return errors.New(E_INTERNAL_ERROR)
 	}
-	for _, row := range txn.cache.Table(txn.request.DBName, *ovsOp.Table) {
-		ok, err := isRowSelectedByWhere(tableSchema, txn.mapUUID, row, ovsOp.Where)
+
+	var equal bool
+	switch *ovsOp.Until {
+	case FN_EQ:
+		equal = true
+	case FN_NE:
+		equal = false
+	default:
+		klog.Errorf("wait: unsupported function %s", *ovsOp.Until)
+		return errors.New(E_CONSTRAINT_VIOLATION)
+	}
+
+	for _, actual := range txn.cache.Table(txn.request.DBName, *ovsOp.Table) {
+		ok, err := isRowSelectedByWhere(tableSchema, txn.mapUUID, actual, ovsOp.Where)
 		if err != nil {
 			return err
 		}
 		if !ok {
 			continue
 		}
-		newRow, err := reduceRowByColumns(row, ovsOp.Columns)
+		actual, err = reduceRowByColumns(actual, ovsOp.Columns)
 		if err != nil {
+			klog.Errorf("wait: failed column reduction %s", err)
 			return err
 		}
-		rows = append(rows, *newRow)
+
+		for _, expected := range *ovsOp.Rows {
+			cond, err := isEqualRow(tableSchema, &expected, actual)
+			if err != nil {
+				klog.Errorf("wait: failed row comapre %s", err)
+				return err
+			}
+			if equal && cond {
+				return nil
+			}
+		}
 	}
 
-	if ovsOp.Rows == nil {
+	if !equal {
```

Comment:
if equal is "false" but the cond  is "true", you return nil, but it should be "time-out" error

Reply:
good catch, fixed